core/extractor/factory.go

```diff
@@ -10,26 +10,26 @@ var (
 )
 
 type Factory struct {
-	fnStore map[string]core.Extractor
+	fnStore map[string]func() core.Extractor
```

Comment:
any reason we are using these functions here instead of directly interface?

Reply:
ok so basically I want this factory to return a new instance of extractor for each run.

reasons are:
1. this would allow extractor to utilise its private fields.
2. and in case an extractor needs to use internal state it should not affect other recipe runs. (esp when we are running multiple recipes asynchronously)

Reply:
it allows me to do this thing [here](https://github.com/odpf/meteor/pull/149/files?file-filters%5B%5D=.go#diff-6157bb7014e75b61c5b682ac904d45bf3d64be94cf695cc1a8c4d40278b7ab0bR30).

I hope by allowing extractor to utilise its internal states, it could lead to better code, esp in readability part.
And by making sure factory always returns a new instance, I dont have to worry about conflicting states.

wdyt?

Reply:
Got it, makes sense. So having an independent internal state for each could be useful but I don't think the example that you gave here is great. We could have just passed it as an argument right?

Reply:
in this case yes I agree, we totally could pass it as an argument and it would be fine, but I'm hoping that could set an example to utilise internal states.

Also on the edge case that the extractor so complex that requires a lot of function callings, we wouldn't have to keep passing the channel everytime.

Reply:
+1 to extractor internal states.